src/main/workflo/macros/screen/om_next.cljs

```diff
@@ -0,0 +1,172 @@
+(ns workflo.macros.screen.om-next
+  (:require [om.next :as om]
+            [workflo.macros.screen.bidi :as sb]
+            [workflo.macros.screen :as scr]
+            [workflo.macros.view :refer-macros [defview]]))
+
+;;;; Remember the active screen
+
+(defonce ^:private +active-screen+
+  (atom {:screen nil :params nil}))
+
+(defn active-screen
+  "Returns the active screen as a map, with :screen storing
+   the current screen and :params storing the URL parameters
+   for the current screen."
+  []
+  @+active-screen+)
+
+(defn set-active-screen!
+  "Sets the active screen and its parameters."
+  [screen params]
+  (swap! +active-screen+ assoc
+         :screen screen
+         :params params))
+
+(defn reload-active-screen!
+  "Reloads the implementation of the active screen by replacing
+   it with a freshly resolved instance from the screen registry."
+  []
+  (swap! +active-screen+ update :screen
+         (fn [screen]
+           (some-> screen :name scr/resolve-screen))))
+
+;;;; Wrapping Om Next parser to handle screen-based routing
+
+(defn read-screen
+  "Executes a query agains the active screen."
+  [read {:keys [parser query] :as env} _ _]
+  {:value (parser (assoc env :screen (:screen (active-screen)))
+                  query nil)})
+
+(defn read-navigation
+  "Queries the navigation info of the active screen."
+  [_ {:keys [screen]} _ _]
+  {:value (:navigation screen)})
+
+(defn read-layout
+  "Executes all queries for views in the layout of the
+   active screen."
+  [_ {:keys [parser query screen] :as env} _ _]
+  {:value (reduce (fn [res layout-item]
+                    (let [[k v] (first layout-item)]
+                      (assoc res k {:view (-> screen :layout
+                                              k :factory)
+                                    :props (parser env v nil)})))
+                  {} query)})
+
+(defn- wrapping-read
+  "Wraps a user-provided parser read function for Om Next by
+   special-casing queries for screen-based routing information
+   (such as :workflo/screen, :workflo/navigation and
+   :workflo/layout.)"
+  [read env key params]
+  (case key
+    :workflo/screen     (read-screen read env key params)
+    :workflo/navigation (read-navigation read env key params)
```

Comment:
We're looking at the root wrapper query here, not the forms in `defscreen`. The way we set this query can be seen in `mount-screen` - at least for `:layout` we need to know that the values of the layout elements refer to views/factories, so we can obtain the right subqueries. We'd have to think more on how we can generalize this further.